// @formio/core/lib/utils/formUtil/__tests__/componentInfo.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const chai_1 = require("chai");
const __1 = require("..");
describe('componentInfo', function () {
    it('should not think table component instance has rows', function () {
        const info = (0, __1.componentInfo)({
            component: {
                label: 'Component',
                cellAlignment: 'left',
                key: 'table',
                type: 'components',
                component: {
                    type: 'table',
                    rows: [
                        [
                            {
                                components: [
                                    {
                                        label: 'Url',
                                        tableView: true,
                                        key: 'url1',
                                        type: 'url',
                                        input: true,
                                    },
                                ],
                            },
                        ],
                    ],
                },
            },
        });
        (0, chai_1.expect)(info.hasRows).to.be.undefined; // eslint-disable-line
    });
});
